src/rpc/inprocess.ts

```diff
@@ -21,6 +21,76 @@ import { PlaywrightDispatcher } from './server/playwrightDispatcher';
 import { setUseApiName } from '../progress';
 import { Connection } from './client/connection';
 import { isUnderTest } from '../helper';
+import { SignalHandler, Closeable } from './client/browserType';
+
+type Signal = 'SIGINT' | 'SIGTERM' | 'SIGHUP';
+const kSignals: Signal[] = ['SIGINT', 'SIGTERM', 'SIGHUP'];
+
+class ProcessSignalHandler implements SignalHandler {
+  private _closeables = new Map<Closeable, Signal[]>();
+  private _listeners = new Map<Signal, (...args: any[]) => void>();
+  private _sigintTimer?: NodeJS.Timeout;
+
+  registerCloseable(closeable: Closeable, options: { handleSIGINT?: boolean, handleSIGHUP?: boolean, handleSIGTERM?: boolean }): void {
+    const {
+      handleSIGINT = true,
+      handleSIGHUP = true,
+      handleSIGTERM = true,
+    } = options;
+    const signals: Signal[] = [];
+    if (handleSIGINT)
+      signals.push('SIGINT');
+    if (handleSIGHUP)
+      signals.push('SIGHUP');
+    if (handleSIGTERM)
+      signals.push('SIGTERM');
+    this._closeables.set(closeable, signals);
+    this._updateListeners();
+  }
+
+  unregisterCloseable(closeable: Closeable): void {
+    this._closeables.delete(closeable);
+    this._updateListeners();
+  }
+
+  private _updateListeners() {
+    for (const signal of kSignals) {
+      const needListener = Array.from(this._closeables.values()).find(signals => signals.includes(signal));
+      let listener = this._listeners.get(signal);
+      if (listener && !needListener) {
+        this._listeners.delete(signal);
+        process.removeListener(signal as any, listener);
+      } else if (needListener && !listener) {
+        listener = () => this._handleSignal(signal);
+        this._listeners.set(signal, listener);
+        process.addListener(signal, listener);
+      }
+    }
+  }
+
+  private async _handleSignal(signal: Signal) {
+    if (signal === 'SIGINT') {
+      // Force exit after 30 seconds.
+      this._sigintTimer = setTimeout(() => process.exit(130), 30000);
+    }
+
+    // Meanwhile, try to gracefully close all browsers.
+    const promises = Array.from(this._closeables.keys()).map(c => c.close());
+    await Promise.all(promises).catch(e => null);
+
+    // Avoid stalling just because of a timer.
+    if (this._sigintTimer)
+      clearTimeout(this._sigintTimer);
+
+    if (signal === 'SIGINT') {
+      // Give tests a chance to dispatch any async calls.
+      if (isUnderTest())
```

Comment:
This looks very sketchy to me. What breaks without it?

Reply:
RPC needs to send some messages so that our test gets a 'close' event so that we can validate that browser did exit with expected signal.